### PR TITLE
internal/blobstore: new package

### PR DESCRIPTION
**Pull request by rogpeppe**
This implements the API that we need for storing
blobs, layered onto juju/blobstore.

### CODE REVIEW

File: internal/multihash/multihash.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+func New(h1, h2 hash.Hash) hash.Hash {
+	return &multiHash{h1, h2}
+}
+
+type multiHash struct {
+	h1, h2 hash.Hash
+}
+
+func (h *multiHash) Write(buf []byte) (int, error) {
+	h.h1.Write(buf)
```
Comment (frankban) [inline internal/multihash/multihash.go:21]: Shouldn't we check errors for each write call here?

File: internal/multihash/multihash.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func (h *multiHash) Size() int {
+	return h.h1.Size() + h.h2.Size()
+}
+
+func (h *multiHash) BlockSize() int {
+	// better: use least common multiple of both block sizes.
+	return h.h1.BlockSize()
+}
```
Comment (frankban) [inline internal/multihash/multihash.go:44]: Having a few tests for the multihash would be great.

**Comment by frankban** [comment]
Thanks for this branch Roger!
:+1: with just some minor suggestions.

File: internal/blobstore/blobstore.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	r, length, err := s.mstore.GetForEnvironment("", hashSum)
+	if err != nil {
+		return nil, 0, err
+	}
+	return r.(ReadSeekCloser), length, nil
+}
+
+// newResourceHash returns a ResourceHash equivalent to the
+// given hashSum. It does not complain if hashSum is invalid - the
+// lower levels will bowl that out.
```
Comment (bac) [inline internal/blobstore/blobstore.go]: Could you rephrase "bowl that out" to be less colloquial?

**Comment by jujugui** [comment]
please test this

File: internal/multihash/multihash.go (hunk 1 above)
Comment (rogpeppe) [inline internal/multihash/multihash.go:21]: No - added a comment.

File: internal/multihash/multihash.go (hunk 2 above)
Comment (rogpeppe) [inline internal/multihash/multihash.go:44]: Done

File: internal/blobstore/blobstore.go (hunk 3 above)
Comment (rogpeppe) [inline internal/blobstore/blobstore.go]: Done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by frankban** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/charmstore-merge/21

**Comment by frankban** [comment]
This time the failure seems to expose a real problem (yay!).

**Comment by rogpeppe** [comment]
Yes - I'd missed adding some dependencies.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge
